common/txmgr/address_state.go

```diff
@@ -224,23 +286,194 @@ func (as *addressState[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) moveUn
 }
 
 // moveTxToFatalError moves a transaction to the fatal error state.
+// It returns an error if there is no transaction with the given ID.
+// It returns an error if the transaction is not in an expected state.
+// Unknown transaction states will cause a panic this includes Unconfirmed and Confirmed transactions.
 func (as *addressState[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) moveTxToFatalError(
 	txID int64, txError null.String,
 ) error {
+	as.Lock()
+	defer as.Unlock()
+
+	tx := as.allTxs[txID]
+	if tx == nil {
+		return fmt.Errorf("move_tx_to_fatal_error: no transaction with ID %d", txID)
+	}
+
+	as._moveTxToFatalError(tx, txError)
+
+	switch tx.State {
+	case TxUnstarted:
+		_ = as.unstartedTxs.RemoveTxByID(txID)
+	case TxInProgress:
+		as.inprogressTx = nil
+	case TxConfirmedMissingReceipt:
+		delete(as.confirmedMissingReceiptTxs, tx.ID)
+	default:
+		panic("move_tx_to_fatal_error: unknown transaction state")
+	}
+
 	return nil
 }
 
 // moveUnconfirmedToConfirmedMissingReceipt moves the unconfirmed transaction to the confirmed missing receipt state.
-func (as *addressState[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) moveUnconfirmedToConfirmedMissingReceipt(attempt txmgrtypes.TxAttempt[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE], broadcastAt time.Time) error {
+// If there is no unconfirmed transaction with the given ID, an error is returned.
+func (as *addressState[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) moveUnconfirmedToConfirmedMissingReceipt(txAttempt txmgrtypes.TxAttempt[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE], broadcastAt time.Time) error {
```

Comment:
My paranoia is getting the better of me here. Should we be wary of accepting the entire txAttempt here and updating our state with this object? If someone inadvertently makes a change to the txAttempt before calling this function, we'd store those changes without validation. It might be safer just accepting the hash and looking it up in our internal map and making the changes on that object.

Reply:
Also I noticed we append the attempt to the tx's attempt list. Wouldn't the attempt already exist in the list if it's an Unconfirmed tx?